src/main/java/com/google/cloud/spanner/pgadapter/utils/MutationWriter.java

```diff
@@ -14,225 +14,622 @@
+
+  private final int commitSizeLimit =
+      Integer.parseInt(
+          System.getProperty("copy_in_commit_limit", String.valueOf(DEFAULT_COMMIT_LIMIT)));
+  private final CopyTransactionMode transactionMode;
+  private final boolean hasHeader;
   private boolean isHeaderParsed;
-  private int mutationCount;
-  private int rowCount;
-  private List<Mutation> mutations;
-  private String tableName;
-  private Map<String, TypeCode> tableColumns;
-  private CSVFormat format;
-  private FileWriter fileWriter;
-  private ByteArrayOutputStream payload = new ByteArrayOutputStream();
+  private long rowCount;
+  private final CloudSpannerJdbcConnection connection;
+  private DatabaseClient databaseClient;
+  private final String tableName;
+  private final Map<String, TypeCode> tableColumns;
+  private final int maxBatchSize;
+  private final long commitSizeLimitForBatching;
+  private final int maxParallelism;
+  private final int pipeBufferSize;
+  private final CSVFormat format;
+  private final CSVParser parser;
+  private PrintWriter errorFileWriter;
+  private final PipedOutputStream payload = new PipedOutputStream();
+  private final AtomicBoolean rollback = new AtomicBoolean(false);
+  private final ListeningExecutorService executorService =
+      MoreExecutors.listeningDecorator(Executors.newCachedThreadPool());
+
+  private final Object lock = new Object();
+
+  @GuardedBy("lock")
+  private SQLException exception;
 
   public MutationWriter(
-      String tableName, Map<String, TypeCode> tableColumns, CSVFormat format, boolean hasHeader) {
-    this.mutationCount = 0;
+      CopyTransactionMode transactionMode,
+      CloudSpannerJdbcConnection connection,
+      String tableName,
+      Map<String, TypeCode> tableColumns,
+      int indexedColumnsCount,
+      CSVFormat format,
+      boolean hasHeader)
+      throws IOException {
+    this(
+        transactionMode,
+        connection,
+        null,
+        tableName,
+        tableColumns,
+        indexedColumnsCount,
+        format,
+        hasHeader);
+  }
+
+  @VisibleForTesting
+  MutationWriter(
+      CopyTransactionMode transactionMode,
+      CloudSpannerJdbcConnection connection,
+      DatabaseClient databaseClient,
+      String tableName,
+      Map<String, TypeCode> tableColumns,
+      int indexedColumnsCount,
+      CSVFormat format,
+      boolean hasHeader)
+      throws IOException {
+    this.transactionMode = transactionMode;
+    this.connection = connection;
+    this.databaseClient = databaseClient;
     this.hasHeader = hasHeader;
     this.isHeaderParsed = false;
     this.tableName = tableName;
     this.tableColumns = tableColumns;
+    int mutationLimit =
+        Math.max(
+            Integer.parseInt(
+                System.getProperty(
+                    "copy_in_mutation_limit", String.valueOf(DEFAULT_MUTATION_LIMIT))),
+            1);
+    this.maxBatchSize = Math.max(mutationLimit / (tableColumns.size() + indexedColumnsCount), 1);
```

Comment:
The 20,000 mutations limit in Spanner is unfortunately a little confusing. It does not mean a maximum of 20,000 mutation objects or rows, it means a maximum of 20,000 column mutations. And it gets a little bit more complex as every secondary index that includes a column that is inserted/updated is also counted as a mutation. So assume that we have the following table and index definition:

```sql
create table numbers (number bigint not null primary key, name varchar(1000));
create index idx_numbers on numbers (name);
```

Any insert into this table will count as 3 mutations:
1. 2 mutations because it the table has two columns.
2. +1 mutation because the `name` column is included in a secondary index.

See also https://cloud.google.com/spanner/quotas#note2